File: server/src/main/java/com/ahmadramadhan/mudahtitip/consignment/Consignment.java
package com.ahmadramadhan.mudahtitip.consignment;

import com.ahmadramadhan.mudahtitip.common.entity.BaseEntity;
import com.ahmadramadhan.mudahtitip.product.Product;
import com.ahmadramadhan.mudahtitip.sale.Sale;
import com.ahmadramadhan.mudahtitip.shop.Shop;
import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Positive;
import jakarta.validation.constraints.PositiveOrZero;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

/**
 * Consignment entity representing a batch of products placed at a shop.
 * Tracks quantity, pricing, commission, expiry, and status.
 */
@Entity
@Table(name = "consignments")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Consignment extends BaseEntity {

    @ManyToOne
    @JoinColumn(name = "product_id", nullable = false)
    private Product product;

    @ManyToOne
    @JoinColumn(name = "shop_id", nullable = false)
    private Shop shop;

    /**
     * Initial quantity placed at the shop.
     */
    @NotNull(message = "Jumlah awal wajib diisi")
    @Positive(message = "Jumlah harus positif")
    @Column(name = "initial_quantity", nullable = false)
    private Integer initialQuantity;

    /**
     * Current remaining quantity (initial - sold - returned).
     */
    @PositiveOrZero(message = "Jumlah tidak boleh negatif")
    @Column(name = "current_quantity", nullable = false)
    private Integer currentQuantity;

    /**
     * Selling price at the shop (may differ from base price).
     */
    @NotNull(message = "Harga jual wajib diisi")
    @Positive(message = "Harga harus positif")
    @Column(name = "selling_price", nullable = false, precision = 12, scale = 2)
    private BigDecimal sellingPrice;

    /**
     * Commission percentage for the shop (e.g., 10 for 10%).
     */
    @NotNull(message = "Persentase komisi wajib diisi")
    @PositiveOrZero(message = "Komisi tidak boleh negatif")
    @Column(name = "commission_percent", nullable = false, precision = 5, scale = 2)
    private BigDecimal commissionPercent;

    @Column(name = "consignment_date")
    private LocalDate consignmentDate;

    @Column(name = "expiry_date")
    private LocalDate expiryDate;

    @Builder.Default
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private ConsignmentStatus status = ConsignmentStatus.ACTIVE;

    @Column(length = 500)
    private String notes;

    /**
     * Sales made from this consignment.
     */
    @JsonIgnore
    @Builder.Default
    @OneToMany(mappedBy = "consignment", cascade = CascadeType.ALL)
    private List<Sale> sales = new ArrayList<>();

    /**
     * Check if consignment is expired based on expiry date.
     */
    public boolean isExpired() {
        if (expiryDate == null)
            return false;
        return LocalDate.now().isAfter(expiryDate);
    }

    /**
     * Check if consignment is expiring within given days.
     */
    public boolean isExpiringWithin(int days) {
        if (expiryDate == null)
            return false;
        return LocalDate.now().plusDays(days).isAfter(expiryDate);
    }
}

File: client/lib/features/consignment/data/models/consignment.dart
import 'package:equatable/equatable.dart';

import '../../../products/data/models/product.dart';

/// Consignment status enum matching backend ConsignmentStatus.
enum ConsignmentStatus {
  active,
  completed,
  expired,
  returned;

  String get displayName => switch (this) {
    ConsignmentStatus.active => 'Aktif',
    ConsignmentStatus.completed => 'Selesai',
    ConsignmentStatus.expired => 'Kadaluarsa',
    ConsignmentStatus.returned => 'Dikembalikan',
  };

  static ConsignmentStatus fromString(String value) {
    return ConsignmentStatus.values.firstWhere(
      (e) => e.name.toUpperCase() == value.toUpperCase(),
      orElse: () => ConsignmentStatus.active,
    );
  }
}

/// Shop model (simplified for consignment context).
class Shop extends Equatable {
  final int id;
  final String name;
  final String? address;
  final String? phone;

  const Shop({required this.id, required this.name, this.address, this.phone});

  factory Shop.fromJson(Map<String, dynamic> json) {
    return Shop(
      id: json['id'] as int,
      name: json['name'] as String,
      address: json['address'] as String?,
      phone: json['phone'] as String?,
    );
  }

  @override
  List<Object?> get props => [id, name, address, phone];
}

/// Consignment model matching backend Consignment entity.
class Consignment extends Equatable {
  final int id;
  final Product product;
  final Shop shop;
  final int initialQuantity;
  final int currentQuantity;
  final double sellingPrice;
  final double commissionPercent;
  final DateTime? consignmentDate;
  final DateTime? expiryDate;
  final ConsignmentStatus status;
  final String? notes;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Consignment({
    required this.id,
    required this.product,
    required this.shop,
    required this.initialQuantity,
    required this.currentQuantity,
    required this.sellingPrice,
    required this.commissionPercent,
    this.consignmentDate,
    this.expiryDate,
    required this.status,
    this.notes,
    required this.createdAt,
    required this.updatedAt,
  });

  /// Quantity that has been sold.
  int get soldQuantity => initialQuantity - currentQuantity;

  /// Check if consignment is expired.
  bool get isExpired =>
      expiryDate != null && DateTime.now().isAfter(expiryDate!);

  /// Check if consignment is expiring within given days.
  bool isExpiringWithin(int days) {
    if (expiryDate == null) return false;
    return DateTime.now().add(Duration(days: days)).isAfter(expiryDate!);
  }

  factory Consignment.fromJson(Map<String, dynamic> json) {
    return Consignment(
      id: json['id'] as int,
      product: Product.fromJson(json['product'] as Map<String, dynamic>),
      shop: Shop.fromJson(json['shop'] as Map<String, dynamic>),
      initialQuantity: json['initialQuantity'] as int,
      currentQuantity: json['currentQuantity'] as int,
      sellingPrice: (json['sellingPrice'] as num).toDouble(),
      commissionPercent: (json['commissionPercent'] as num).toDouble(),
      consignmentDate: json['consignmentDate'] != null
          ? DateTime.parse(json['consignmentDate'] as String)
          : null,
      expiryDate: json['expiryDate'] != null
          ? DateTime.parse(json['expiryDate'] as String)
          : null,
      status: ConsignmentStatus.fromString(json['status'] as String),
      notes: json['notes'] as String?,
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
    );
  }

  @override
  List<Object?> get props => [
    id,
    product,
    shop,
    initialQuantity,
    currentQuantity,
    sellingPrice,
    commissionPercent,
    consignmentDate,
    expiryDate,
    status,
    notes,
    createdAt,
    updatedAt,
  ];
}
